integration-tests/src/integration_test_circuits.rs

```diff
@@ -0,0 +1,332 @@
+use crate::{get_client, GenDataOutput, CHAIN_ID};
+use bus_mapping::circuit_input_builder::{BuilderClient, CircuitInputBuilder, CircuitsParams};
+use eth_types::geth_types;
+use halo2_proofs::plonk::{
+    create_proof, keygen_pk, keygen_vk, verify_proof, Circuit, ProvingKey, VerifyingKey,
+};
+use halo2_proofs::poly::commitment::ParamsProver;
+use halo2_proofs::poly::kzg::commitment::{KZGCommitmentScheme, ParamsKZG, ParamsVerifierKZG};
+use halo2_proofs::poly::kzg::multiopen::{ProverSHPLONK, VerifierSHPLONK};
+use halo2_proofs::poly::kzg::strategy::SingleStrategy;
+use halo2_proofs::{
+    arithmetic::CurveAffine,
+    dev::MockProver,
+    halo2curves::{
+        bn256::Fr,
+        group::{Curve, Group},
+    },
+};
+use halo2_proofs::{
+    halo2curves::bn256::{Bn256, G1Affine},
+    transcript::{
+        Blake2bRead, Blake2bWrite, Challenge255, TranscriptReadBuffer, TranscriptWriterBuffer,
+    },
+};
+use lazy_static::lazy_static;
+use log::trace;
+use rand_chacha::rand_core::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+use rand_core::RngCore;
+use rand_xorshift::XorShiftRng;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+use zkevm_circuits::bytecode_circuit::bytecode_unroller::unroll;
+use zkevm_circuits::bytecode_circuit::dev::BytecodeCircuitTester;
+use zkevm_circuits::copy_circuit::dev::CopyCircuitTester;
+use zkevm_circuits::evm_circuit::test::{get_test_degree, get_test_instance};
+use zkevm_circuits::evm_circuit::witness::RwMap;
+use zkevm_circuits::evm_circuit::{test::get_test_cicuit_from_block, witness::block_convert};
+use zkevm_circuits::state_circuit::StateCircuit;
+use zkevm_circuits::super_circuit::SuperCircuit;
+use zkevm_circuits::tx_circuit::{sign_verify::SignVerifyChip, Secp256k1Affine, TxCircuit};
+
+lazy_static! {
+    /// Data generation.
+    pub static ref GEN_DATA: GenDataOutput = GenDataOutput::load();
+    static ref RNG: XorShiftRng = XorShiftRng::from_seed([
+        0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06, 0xbc,
+        0xe5,
+    ]);
+    static ref GEN_PARAMS: Mutex<HashMap<u32, ParamsKZG<Bn256>>> = Mutex::new(HashMap::new());
+}
+
+const CIRCUITS_PARAMS: CircuitsParams = CircuitsParams {
+    max_rws: 16384,
+    max_txs: 4,
+    keccak_padding: None,
+};
+
+fn get_general_params(degree: u32) -> ParamsKZG<Bn256> {
+    match GEN_PARAMS.lock().unwrap().get(&degree) {
+        Some(params) => params.clone(),
+        None => {
+            let params = ParamsKZG::<Bn256>::setup(degree, RNG.clone());
+            GEN_PARAMS.lock().unwrap().insert(degree, params.clone());
```

Comment:
This looks like a double lock on `GEN_PARAMS`.  Did you test this successfully?

Reply:
Yes, in some way this is a oversight from my part as I should have seen it. The fact of adding Mutex is just an easy way to allow a mutable lazy_init, so it was artificial and I didn't look enough into it.

However, it works, because I think rust drops it as an optiomisation when the lock is not used anymore right after the lock. But as I saw in other code that is not always the same and we should not relay on this. There were a ticket open by pinkie about this.

I will make it lock just once.